src/Evolution/Systems/GeneralizedHarmonic/Equations.cpp

```diff
@@ -307,6 +333,178 @@ void ComputeNormalDotFluxes<Dim>::apply(
+  get<Tags::ConstraintGamma1>(*packaged_data) = gamma1;
+  get<Tags::ConstraintGamma2>(*packaged_data) = gamma2;
+  get<::Tags::UnitFaceNormal<Dim, Frame::Inertial>>(*packaged_data) =
+      interface_unit_normal;
+}
+
+template <size_t Dim>
+void UpwindFlux<Dim>::operator()(
+    gsl::not_null<db::item_type<::Tags::NormalDotNumericalFlux<
+        gr::Tags::SpacetimeMetric<Dim, Frame::Inertial, DataVector>>>*>
+        psi_normal_dot_numerical_flux,
+    gsl::not_null<db::item_type<::Tags::NormalDotNumericalFlux<
+        GeneralizedHarmonic::Tags::Pi<Dim, Frame::Inertial>>>*>
+        pi_normal_dot_numerical_flux,
+    gsl::not_null<db::item_type<::Tags::NormalDotNumericalFlux<
+        GeneralizedHarmonic::Tags::Phi<Dim, Frame::Inertial>>>*>
+        phi_normal_dot_numerical_flux,
+    const db::item_type<
+        gr::Tags::SpacetimeMetric<Dim, Frame::Inertial, DataVector>>&
+        spacetime_metric_int,
+    const db::item_type<Tags::Pi<Dim, Frame::Inertial>>& pi_int,
+    const db::item_type<Tags::Phi<Dim, Frame::Inertial>>& phi_int,
+    const db::item_type<gr::Tags::Lapse<DataVector>>& lapse_int,
+    const db::item_type<gr::Tags::Shift<Dim, Frame::Inertial, DataVector>>&
+        shift_int,
+    const db::item_type<
+        gr::Tags::InverseSpatialMetric<Dim, Frame::Inertial, DataVector>>&
+        inverse_spatial_metric_int,
+    const db::item_type<Tags::ConstraintGamma2>& gamma1_int,
+    const db::item_type<Tags::ConstraintGamma2>& gamma2_int,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal_int,
+    const db::item_type<
+        gr::Tags::SpacetimeMetric<Dim, Frame::Inertial, DataVector>>&
+        spacetime_metric_ext,
+    const db::item_type<Tags::Pi<Dim, Frame::Inertial>>& pi_ext,
+    const db::item_type<Tags::Phi<Dim, Frame::Inertial>>& phi_ext,
+    const db::item_type<gr::Tags::Lapse<DataVector>>& lapse_ext,
+    const db::item_type<gr::Tags::Shift<Dim, Frame::Inertial, DataVector>>&
+        shift_ext,
+    const db::item_type<
+        gr::Tags::InverseSpatialMetric<Dim, Frame::Inertial, DataVector>>&
+        inverse_spatial_metric_ext,
+    const db::item_type<Tags::ConstraintGamma2>& gamma1_ext,
+    const db::item_type<Tags::ConstraintGamma2>& gamma2_ext,
+    const tnsr::i<DataVector, Dim,
+                  Frame::Inertial>& /*interface_unit_normal_ext*/) const
+    noexcept {
+  const Scalar<DataVector> gamma1_avg{0.5 *
+                                      (get(gamma1_int) + get(gamma1_ext))};
+  const Scalar<DataVector> gamma2_avg{0.5 *
+                                      (get(gamma2_int) + get(gamma2_ext))};
+
+  const auto& char_fields_int = characteristic_fields(
+      gamma2_avg, inverse_spatial_metric_int, spacetime_metric_int, pi_int,
+      phi_int, interface_unit_normal_int);
+  const auto& char_speeds_int = characteristic_speeds(
+      gamma1_avg, lapse_int, shift_int, interface_unit_normal_int);
+  const auto& char_fields_ext = characteristic_fields(
```

Comment:
No. That turns out to be unstable. The char fields on an element are always defined in terms of the normal of that element. The upwind flux only says that when you compute incoming char fields, use the evolved fields from the neighbor; for all other char fields, use quantities in the element.

Reply:
Won't that violate the requirement that the two elements calculate the same numerical flux?

Reply:
I don’t think so. Empirically, as codes here is stable, but using the other normal is not. Physically, I think you’re doing the same thing on both sides: use neighbor incoming char fields, self outgoing char fields.

Reply:
I'm asking about correctness, not stability.

I was going to try to check this by modifying your test, but it looks like the test never actually calls anything except the internal `weight_char_fields` function.  Please modify your test so that it tests the public interface of your class.